### ISSUE TITLE
(Maybe) spurious "possible unbound" report

### ISSUE THREAD

**Issue by wbadart** [microsoft/pyright#1507]
**Describe the bug**

Pyright reports a possibly unbound variable on the right-hand side of an `or` expression which will only ever be evaluated for code-paths where the variable is bound (due to short-circuiting).

**To Reproduce**

```python
try:
    x = something_that_might_blow_up()
    blew_up = False
except:
    blew_up = True
if blew_up or some_other_condition(x):  # <-- reports 'x' as possibly unbound
    raise ValueError("my odd check has failed")
```

**Expected behavior**

I know this is a somewhat perverse use of control flow, so I didn't fully expect it to be picked up, but I thought that if I could convince myself that the above snippet is safe, Pyright might be able to detect the same.

**VS Code extension or command-line**

Using `v1.1.109` in neovim via [`coc-pyright`](https://github.com/fannheyward/coc-pyright).

Thanks for your consideration!

**Comment by erictraut** [comment]
This would require an understanding of relationships between two variables — `x` and `blew_up`. In particular, it would have to know that the type of `x` is unbound only in situations where the type of `blew_up` is `Literal[True]`.

Pyright (and all type checkers I'm aware of for all languages) are not able to do anything that sophisticated, and we don't have any plans to add such support. It would likely be computationally intractable. Tracking types for individual variables independently of each other is already a challenging task. 

The good news is that there's an easy workaround. Simply initialize `x` prior to the `try` block with a sentinel value like `None`.

```python
x = None
try:
    x = something_that_might_blow_up()
except:
    pass
if x is not None and some_other_condition(x):  # <-- reports 'x' as possibly unbound
    raise ValueError("my odd check has failed")
```

**Comment by wbadart** [comment]
Thanks for the response, makes sense. I'll go with the sentinel approach 👍
